### PR TITLE
win32: take dpi scale into account with [current-]window-scale

### PR DESCRIPTION
**Pull request by avih**
win32: take dpi scale into account with [current-]window-scale (osx/x11/wayland already do that), and two minor refinements:
- Ensure initial dpi-scale value.
- Support the property `display-hidpi-scale` (only used from `console.lua`).

### CODE REVIEW

File: video/out/w32_common.c [hunk 1]
```diff
@@ -555,6 +555,10 @@ static void update_dpi(struct vo_w32_state *w32)
         w32->dpi_scale = 1.0;
         MP_VERBOSE(w32, "Couldn't determine DPI, falling back to %d\n", w32->dpi);
     }
+
+done:
+    if (w32->dpi_scale <= 0)
```
[the remark this answers is not in the document]
Comment (jeeb) [reply video/out/w32_common.c:560]: Also I do wonder how likely it is that after [`GetDpiForMonitor`](https://docs.microsoft.com/en-us/windows/win32/api/shellscalingapi/nf-shellscalingapi-getdpiformonitor) returns `S_OK`, you get garbage values... It seems like Windows moved from plainly querying from drivers to having its own calculation logic with the WDDM (via [here](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-getdevicecaps)):

> Unfortunately, a display driver that is implemented to the Windows Display Driver Model (WDDM) (introduced in Windows Vista) causes GDI to not get the info, so GetDeviceCaps must always calculate the info.

Basically just tired and wondering if the goto'd forced checking is really required or not :P (and whether you have any real suspicions with regards to `GetDeviceCaps` or so, or if this is just forcing a check on everything).

**Review by jeeb — CHANGES_REQUESTED** [review]
In general looks good to me, just some notes on `update_dpi` and something regarding visual positioning with the multiplication.

Since dpi_scale does not get changed in case of `hidpi_window_scale` being a falsie, this /should/ also not really break things as I do not see `VOCTRL_SET_UNFS_WINDOW_SIZE` being commanded in unrelated places.

File: video/out/w32_common.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (avih) [reply video/out/w32_common.c:560]: > Also I do wonder how likely it is that after [`GetDpiForMonitor`](https://docs.microsoft.com/en-us/windows/win32/api/shellscalingapi/nf-shellscalingapi-getdpiformonitor) returns `S_OK`, you get garbage values...

> Basically just tired and wondering if the goto'd forced checking is really required or not

I have no reason to suspect garbage values. However, the following commit divides by this value, so it's a safety measure to ensure that no matter what - we never divide by 0.

> If dpi_scale ended up invalid then dpi from which it was calculated most likely ended up such as well. If this is done, the logic would look pretty similar to the else case just before.

You mean adding something like this?
```c
if (w32->dpi <= 0)
    w32->dpi = 96;
```

I'd be fine with that.

File: video/out/w32_common.c [hunk 2]
```diff
@@ -1740,6 +1740,8 @@ static int gui_thread_control(struct vo_w32_state *w32, int request, void *arg)
             return VO_FALSE;
 
         RECT *rc = w32->current_fs ? &w32->prev_windowrc : &w32->windowrc;
+        s[0] *= w32->dpi_scale;
```
[the remark this answers is not in the document]
Comment (avih) [reply video/out/w32_common.c:1747]: Sure. Will do.

**Comment by avih** [comment]
Addressed both comments, as well as additional comments given on IRC (at the first commit - refactor `update_dpi()` to remove duplications and improve clarity).

Thanks for the review.

**Comment by ValZapod** [comment]
Do you know how to overwrite "DPI detected from the new API" for HDMI display in mpv and systemwide?? My LG C9 of 55 reports 77' in EDID. Crazy.
